src/passes/TypeGeneralizing.cpp

```diff
@@ -0,0 +1,438 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "analysis/cfg.h"
+#include "analysis/lattice.h"
+#include "analysis/lattices/inverted.h"
+#include "analysis/lattices/shared.h"
+#include "analysis/lattices/stack.h"
+#include "analysis/lattices/tuple.h"
+#include "analysis/lattices/valtype.h"
+#include "analysis/lattices/vector.h"
+#include "analysis/monotone-analyzer.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "wasm-traversal.h"
+#include "wasm.h"
+
+#define TYPE_GENERALIZING_DEBUG 0
+
+#if TYPE_GENERALIZING_DEBUG
+#define DBG(statement) statement
+#else
+#define DBG(statement)
+#endif
+
+// Generalize the types of program locations as much as possible, both to
+// eliminate unnecessarily refined types from the type section and (TODO) to
+// weaken casts that cast to unnecessarily refined types. If the casts are
+// weakened enough, they will be able to be removed by OptimizeInstructions.
+//
+// Perform a backward analysis tracking requirements on the types of program
+// locations (currently just locals and stack values) to discover how much the
+// type of each location can be generalized without breaking validation or
+// changing program behavior.
+
+namespace wasm {
+
+namespace {
+
+using namespace analysis;
+
+// We will learn stricter and stricter requirements as we perform the analysis,
+// so more specific types need to be higher up the lattice.
+using TypeRequirement = Inverted<ValType>;
+
+// Record a type requirement for each local variable. Shared the requirements
+// across basic blocks.
+using LocalTypeRequirements = Shared<Vector<TypeRequirement>>;
+
+// The type requirements for each reference-typed value on the stack at a
+// particular location.
+using ValueStackTypeRequirements = Stack<TypeRequirement>;
+
+// The full lattice used for the analysis.
+using StateLattice =
+  analysis::Tuple<LocalTypeRequirements, ValueStackTypeRequirements>;
+
+// Equip the state lattice with helpful accessors.
+struct State : StateLattice {
+  using Element = StateLattice::Element;
+
+  static constexpr int LocalsIndex = 0;
+  static constexpr int StackIndex = 1;
+
+  State(Function* func) : StateLattice{Shared{initLocals(func)}, initStack()} {}
+
+  void push(Element& elem, Type type) const noexcept {
+    stackLattice().push(stack(elem), std::move(type));
```

Comment:
This should behave the same as copying it, but we require an rvalue reference here so we need to use `std::move`. The rationale is that if moving were different from copying, we would only want to allow moves to avoid expensive implicit copies.